src/udiskslinuxblock.c

```diff
@@ -949,6 +949,64 @@ udisks_linux_find_child_configuration (UDisksDaemon *daemon,
 
 /* ---------------------------------------------------------------------------------------------------- */
 
+/**
+ * Determines whether a block device seems to be encrypted.
+ *
+ * TCRYPT volumes are not easily identifiable, because they have no
+ * cleartext header, but are completely encrypted. This function is
+ * used to determine whether a block device is a candidate for being
+ * TCRYPT encrypted.
+ *
+ * To achieve this, we calculate the chi square value of the first
+ * 512 Bytes and treat devices with a chi square value between 136
+ * and 426 as candidates for being encrypted.
+ * For the reasoning, see: https://tails.boum.org/blueprint/veracrypt/
+ */
+
+#define CHI_SQUARE_LOWER_LIMIT 136
+#define CHI_SQUARE_UPPER_LIMIT 426
+#define BYTES_TO_CHECK 512
+
+static gboolean
+block_seems_encrypted (UDisksBlock *block)
+{
+  const gchar *device = NULL;
+  gint fd = -1;
+  guchar buf[BYTES_TO_CHECK];
+  guint symbols[256] = {0};
+  gfloat chi_square = 0.0;
+  gfloat e = (gfloat) sizeof(buf) / (gfloat) 256.0;
+  guint i;
+
+  device = udisks_block_get_device (block);
+
+  fd = open (device, O_RDONLY);
+  if (fd == -1) {
+      udisks_info("Can't determine if device '%s' seems to be encrypted: Failed to open", device);
+      return FALSE;
+    }
+
+  if (read (fd, buf, sizeof(buf)) != sizeof(buf)) {
+      udisks_info("Can't determine if device '%s' seems to be encrypted: Failed to read", device);
+      close(fd);
+      return FALSE;
+    }
+
+  close(fd);
+
+  /* Calculate Chi Square */
+  for (i = 0; i < sizeof(buf); i++)
+    /* This is safe because the max value of buf[i] is < sizeof(symbols). */
+    symbols[buf[i]]++;
+  for (i = 0; i < 256; i++)
+    chi_square += (symbols[i] - e) * (symbols[i] - e);
+  chi_square /= e;
+
+  return CHI_SQUARE_LOWER_LIMIT < chi_square && chi_square < CHI_SQUARE_UPPER_LIMIT;
+}
```

Comment:
I don't have enough time to look at all the changes here, but I can say right away that this function belongs to *libblockdev*.

Reply:
Makes sense. Done in commit 3d4a065 and bb3c036 in *libblockdev*.